{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4dc60bbb-c7c5-4908-bf19-ae82a8a8fb02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Modules and Packages for Library Generation\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import Bio.Data.CodonTable as CodonTable\n",
    "from itertools import product\n",
    "from dnachisel import *\n",
    "\n",
    "# Define Functions\n",
    "\n",
    "def csv_to_fasta(csv_file, fasta_file, seq_column):\n",
    "    df = pd.read_csv(csv_file)\n",
    "    with open(fasta_file, 'w') as f:\n",
    "        for index, row in df.iterrows():\n",
    "            f.write(f\">{row[0]}\\n{row[seq_column]}\\n\")\n",
    "\n",
    "def initial_reverse_translate(protein_seq):\n",
    "    codon_table = CodonTable.standard_dna_table.forward_table\n",
    "    dna_seq = \"\"\n",
    "    for aa in protein_seq:\n",
    "        codons = [codon for codon, amino_acid in codon_table.items() if amino_acid == aa]\n",
    "        if codons:\n",
    "            dna_seq += codons[0]\n",
    "        else:\n",
    "            raise ValueError(f\"No codon found for amino acid: {aa}\")\n",
    "\n",
    "    return dna_seq\n",
    "\n",
    "def standard_dna_chisel(dna_seq, restriction_enzymes_to_silence):\n",
    "\n",
    "    constraints = [\n",
    "        AvoidHairpins(stem_size=20, hairpin_window=200),\n",
    "        EnforceGCContent(mini=0.30, maxi=0.70, window=50),\n",
    "        EnforceTranslation(location=(0, len(dna_seq)))\n",
    "    ]\n",
    "\n",
    "    for pattern in restriction_enzymes_to_silence:\n",
    "        constraints.append(AvoidPattern(pattern))\n",
    "\n",
    "    problem = DnaOptimizationProblem(\n",
    "        sequence=dna_seq,\n",
    "        constraints=constraints,\n",
    "        objectives=[\n",
    "            CodonOptimize(species='e_coli', method=\"match_codon_usage\", location=(0, len(dna_seq)))\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    problem.resolve_constraints()\n",
    "    problem.optimize()\n",
    "    problem.max_random_iters = 5000\n",
    "\n",
    "    return problem.sequence\n",
    "\n",
    "def barcode_builder(barcode_length, restriction_enzymes, restriction_enzymes_to_silence):\n",
    "    nucleotides = ['A', 'T', 'C', 'G']\n",
    "    barcodes = [''.join(combo) for combo in product(nucleotides, repeat=barcode_length)]\n",
    "\n",
    "    restriction_sites_to_remove = set()\n",
    "    for enzyme in restriction_enzymes_to_silence:\n",
    "        restriction_sites_to_remove.update(restriction_enzymes.get(enzyme, []))\n",
    "\n",
    "    filtered_barcodes = [bc for bc in barcodes if bc not in restriction_sites_to_remove]\n",
    "\n",
    "    return filtered_barcodes\n",
    "\n",
    "def read_fasta_as_list(fasta_file):\n",
    "    sequences = [str(record.seq) for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
    "    return sequences\n",
    "\n",
    "def gc_content(sequence):\n",
    "    sequence = sequence.upper()\n",
    "    gc_count = sequence.count('G') + sequence.count('C')\n",
    "    total_length = len(sequence)\n",
    "\n",
    "    if total_length == 0:\n",
    "        return 0.0\n",
    "\n",
    "    return (gc_count / total_length) * 100\n",
    "\n",
    "def retranslate(dna_seq, start_pos):\n",
    "    dna_seq = dna_seq.upper()\n",
    "    coding_region = dna_seq[start_pos:]\n",
    "    protein_seq = Seq(coding_region).translate(to_stop=True)\n",
    "    \n",
    "    return str(protein_seq)\n",
    "\n",
    "def write_sequences_to_fasta(sequences, output_file, header_prefix=\"Seq\"):\n",
    "    records = []\n",
    "    for i, seq in enumerate(sequences):\n",
    "        record = SeqRecord(Seq(seq), id=f\"{header_prefix}_{i+1}\", description=\"\")\n",
    "        records.append(record)\n",
    "    \n",
    "    with open(output_file, \"w\") as f:\n",
    "        SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "# Define useful variables\n",
    "\n",
    "restriction_enzymes = {\n",
    "    \"BsmBI\" : [\"CGTCTC\",\"GAGACG\"],\n",
    "    \"BsaI\" : [\"GGTCTC\",\"GAGACC\"],\n",
    "    \"BbsI\" : [\"GAAGAC\",\"GTCTTC\"],\n",
    "    \"SapI\" : [\"GCTCTTC\",\"GAAGAGC\"],\n",
    "    \"PaqCI\" : [\"CACCTGC\",\"GCAGGTG\"]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bd62449a-b089-4ee3-89f0-693ce28dd4aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define Pertinent Variables\n",
    "\n",
    "### Filepath to where the CSV full of designs is -- Designs should be AA sequences\n",
    "designs_csv_path = \"/Users/asmiley/Downloads/design_sequences_tier1.csv\"\n",
    "\n",
    "### Define the filepath and filename for a the .FASTA version of the CSV that will be made\n",
    "designs_fasta_path = \"/Users/asmiley/Downloads/design_sequences_tier1.fasta\"\n",
    "\n",
    "### Define the length of a barcode to append to your sequence (4^N = Barcode Diversity)\n",
    "barcode_length = 6\n",
    "\n",
    "### Define the restriction enzymes you want to silence in your library as a list of strings \n",
    "### Supported restriction enzymes -> \"BsmBI\", \"BsaI\", \"BbsI\", \"SapI\", and \"PaqCI\"\n",
    "restriction_enzymes_to_silence = [\"BbsI\", \"PaqCI\"]\n",
    "\n",
    "### Define your favorite stop codon\n",
    "stop_codon = \"TAA\"\n",
    "\n",
    "### Define filepaths for output files\n",
    "output_csv = \"/Users/asmiley/Downloads/test_output.csv\"\n",
    "output_csv_sequences_only = \"/Users/asmiley/Downloads/test_output_sequences_only.csv\"\n",
    "output_fasta_sequences_only = \"/Users/asmiley/Downloads/test_output_sequences_only.fasta\"\n",
    "\n",
    "### Define any consistent sequences you want to be common to all DNA sequences\n",
    "### See the final cell on how to implement this\n",
    "five_prime_constant = \"TTAGTATATTAGTTAAGTATAAGAAGGAGATATACATGATAGAAGCACCAGATGTTAAACCTTGGCTATTCTTG\"\n",
    "three_prime_constant = \"\"\n",
    "internal_primer_binding_site = \"\"\n",
    "five_prime_primer_binding_site = \"\"\n",
    "three_prime_primer_binding_site = \"\"\n",
    "\n",
    "### Define any golden gate sites that you want to implement on your sequences\n",
    "### Do NOT include these if you are using OMEGA oligos\n",
    "### These should include 5 nt of spacer sequence, the cleavage motif, and the overhang\n",
    "five_prime_golden_gate = \"\"\n",
    "three_prime_golden_gate = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acbd3843-aa28-4f75-887d-e1b662b8e97c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DNA Chisel\n",
    "\n",
    "### Converts Design CSV to .FASTA\n",
    "csv_to_fasta(designs_csv_path, designs_fasta_path, seq_column='design_sequence')\n",
    "\n",
    "# Read in .FASTA as a list for optimization and restrict it to the variable section if necessary\n",
    "list_of_designs = read_fasta_as_list(designs_fasta_path)\n",
    "pertinent_sections = [sequence[30:] for sequence in list_of_designs]\n",
    "\n",
    "# Make barcodes, ensuring the removal of any desired restriction sites\n",
    "barcodes = barcode_builder(barcode_length, restriction_enzymes, restriction_enzymes_to_silence)\n",
    "\n",
    "# DNA chisel on the sequences of interest\n",
    "optimized_sequences = [standard_dna_chisel(initial_reverse_translate(sequence), restriction_enzymes_to_silence) for sequence in pertinent_sections]\n",
    "print(\"DNA Chisel Complete\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7a79b0ed-b455-4f4a-bb90-a0ac1bb21adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# File Export\n",
    "\n",
    "columns = ['De Novo AA Sequence', 'Re-Translation of DNA Sequence', 'DNA Sequence', 'Barcode Sequence', \"Length (bp)\", \"% GC Content\"]\n",
    "df = pd.DataFrame(columns=columns)\n",
    "\n",
    "for sequence_idx in range(len(optimized_sequences)):\n",
    "    \n",
    "    # This is where you define your overall sequences with constant sections\n",
    "    optimized_full_length_sequence =  five_prime_golden_gate + barcodes[sequence_idx] + five_prime_constant + optimized_sequences[sequence_idx] + stop_codon + three_prime_golden_gate\n",
    "    \n",
    "    df.loc[sequence_idx] = [pertinent_sections[sequence_idx], retranslate(optimized_full_length_sequence, 41), optimized_full_length_sequence, barcodes[sequence_idx], len(optimized_full_length_sequence), gc_content(optimized_full_length_sequence)]\n",
    "\n",
    "df.to_csv(output_csv)\n",
    "\n",
    "seqs_only = df[['DNA Sequence']]\n",
    "\n",
    "seqs_only.to_csv(output_csv_sequences_only)\n",
    "\n",
    "seqs_only_list = df['DNA Sequence'].tolist()\n",
    "\n",
    "write_sequences_to_fasta(seqs_only_list, output_fasta_sequences_only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a633ad7e-567c-45dd-92e2-1f9a37241d63",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
